lxd/identities.go

```diff
@@ -166,6 +174,358 @@ func identityAccessHandler(authenticationMethod string, entitlement auth.Entitle
+		})
+		if err != nil {
+			return response.InternalError(fmt.Errorf("Failed to create pending TLS identity: %w", err))
+		}
+
+		token := api.CertificateAddToken{
+			ClientName:  req.Name,
+			Fingerprint: fingerprint,
+			Addresses:   addresses,
+			Secret:      joinSecret,
+			ExpiresAt:   expiresAt,
+			Type:        api.IdentityTypeCertificateClient,
+		}
+
+		lc, err := notify(lifecycle.IdentityCreated, identifier.String(), false)
+		if err != nil {
+			return response.SmartError(err)
+		}
+
+		return response.SyncResponseLocation(true, token, lc.Source)
+	}
+
+	if !trusted {
+		// Check if certificate add token supplied as token.
+		joinToken, err := shared.CertificateTokenDecode(req.TrustToken)
+		if err != nil {
+			return response.Forbidden(nil)
+		}
+
+		// If so then check there is a matching pending TLS identity.
+		identifier, err := tlsIdentityTokenValid(r.Context(), s, *joinToken)
+		if err != nil {
+			return response.InternalError(fmt.Errorf("Failed during search for pending TLS identity: %w", err))
+		}
+
+		err = s.DB.Cluster.Transaction(r.Context(), func(ctx context.Context, tx *db.ClusterTx) error {
+			return dbCluster.ActivateTLSIdentity(ctx, tx.Tx(), identifier, cert)
+		})
+		if err != nil {
+			return response.SmartError(err)
+		}
+
+		lc, err := notify(lifecycle.IdentityUpdated, identifier.String(), true)
+		if err != nil {
+			return response.SmartError(err)
+		}
+
+		return response.SyncResponseLocation(true, nil, lc.Source)
+	}
+
+	// Calculate the fingerprint.
+	fingerprint := shared.CertFingerprint(cert)
+	metadata := dbCluster.CertificateMetadata{Certificate: string(pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: cert.Raw}))}
+	metadataJSON, err := json.Marshal(metadata)
+	if err != nil {
+		return response.InternalError(fmt.Errorf("Failed to encode certificate metadata: %w", err))
+	}
+
+	err = s.DB.Cluster.Transaction(context.TODO(), func(ctx context.Context, tx *db.ClusterTx) error {
+		// Check if we already have the certificate.
+		_, err := dbCluster.GetIdentityID(ctx, tx.Tx(), api.AuthenticationMethodTLS, fingerprint)
+		if err == nil {
+			return api.StatusErrorf(http.StatusConflict, "Identity already exists")
+		}
+
+		id, err := dbCluster.CreateIdentity(ctx, tx.Tx(), dbCluster.Identity{
+			AuthMethod: api.AuthenticationMethodTLS,
+			Type:       api.IdentityTypeCertificateClient,
+			Identifier: fingerprint,
+			Name:       req.Name,
+			Metadata:   string(metadataJSON),
+		})
+		if err != nil {
+			return err
+		}
+
+		return dbCluster.SetIdentityAuthGroups(ctx, tx.Tx(), int(id), req.Groups)
+	})
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	lc, err := notify(lifecycle.IdentityCreated, fingerprint, true)
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	return response.SyncResponseLocation(true, nil, lc.Source)
+}
+
+func tlsIdentityTokenValid(ctx context.Context, s *state.State, token api.CertificateAddToken) (uuid.UUID, error) {
```

Comment:
could this end "Validate" so it aligns consistent with `certificateValidate` above?